docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,218 @@
+- Provide a mechanisim to guarantee that application quorum for N members is maintained at any time.
+
+## Proposal
+The machine health checker (MHC) does a best effort to keep nodes healthy in the cluster, by removing machines backing unhealthy Nodes.
+
+It provides a short-circuit mechanism and limits remediation when the `maxUnhealthy` threshold is reached for a targeted group of machines.
+This is similar to what the node life cycle controller does for reducing the eviction rate as nodes become unhealthy in a given zone. E.g a large number of nodes in a single zone are down due to a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+### Unhealthy criteria:
+A machine is unhealthy when:
+- The referenced node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has a nodeRef but the referenced node is not found.
+- The Machine is in phase "Failed", It has `ErrorMessage`/`ErrorReason` set.
+
+If any of those criteria are met for longer than the given timeouts and the `maxUnhealthy` threshold has not been reached yet, remediation is triggered.
+
+Timeouts:
+- For the node conditions the time outs are defined by the admin.
+- For a machine with no nodeRef an opinionated value could be assumed e.g 10 min.
+- For a node notFound or a failed machine, the machine is considerable unrecoverable, remediation can be triggered right away if `maxUnhealthy` is not reached.
+
+### Remediation:
+- A deletion request for the machine is sent to the API server.
+- The controller owning that `Machine`, e.g machineSet reconciles towards the expected number of replicas and start the process to bring up a new `Machine`.
+- The machine controller drains the node.
+- The machine controller and the provider implementation delete the cloud instance.
+- The machine controller removes the finalizer for the `Machine` resource.
+
+
+### User Stories
+
+#### Story 1
+- As a user of a Workload Cluster, I only care about my app's availability, so I want my cluster infrastructure to be self-healing and the nodes to be remediated transparently in the case of failures.
+
+#### Story 2
+As an operator of a Management Cluster, I want my machines to be self-healing and to be recreated, resulting in a new healthy node in the case of matching my unhealthy criteria.
+
+### Implementation Details/Notes/Constraints
+
+#### MachineHealthCheck CRD:
+- Enable watching a group of machines (based on a label selector).
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further remediation will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
```

Comment:
This seems like it could easily be misconfigured.  We should look at the total number of unhealthy nodes, not just the nodes in the 'target pool'.  Also, we're watching a group of machines, not a group of nodes.  Also, it's unclear what should happen if a node/machine is covered in multiple 'groups of machines' mentioned above.

Reply:
I'm not sure I agree here. I would expect users to likely have groups of multiple MachineDeployments and to leverage those as pools of Nodes with different scheduling requirements (GPU availability, public/private app, etc), and as such I would expect to be able to define pertinent health checks against each of these separately.

Reply:
What I'm talking about is a cluster-wide view of unhealthy nodes rather than a view of just a subset of the nodes when making the determination of whether or not to remediate problematic nodes.

Also, it's still unclear what should happen if a node is in multiple groups.

Reply:
> What I'm talking about is a cluster-wide view of unhealthy nodes rather than a view of just a subset of the nodes when making the determination of whether or not to remediate problematic nodes.

Yes, I understand that. I'm saying that for non-trivial use cases (i.e. one MachineDeployment for all worker nodes) I think this is not the view that we care about. We care more about the interruption of the pool of nodes that are serving individual scheduling concerns rather than the full set of nodes in the cluster.

I suspect this is something that we will need to make sure has plenty of clear documentation around, though.

> Also, it's still unclear what should happen if a node is in multiple groups.

This is probably something that should be clarified.

Reply:
>Also, it's still unclear what should happen if a node is in multiple groups.

A node that happens to be covered by more than one MHC is liable to be remediated by any of them satisfying the requirements.

Later on we could discuss things like rejecting multiple groups, rate limiting instead of fully short circuiting, may be add zone awareness and cluster wide view, etc. but I'd prefer to keep this proposal simple and follow up with RFEs as we start gathering feedback based on an initial tangible implementation.